keps/sig-etcd/4331-livez-readyz/README.md

```diff
@@ -0,0 +1,116 @@
+| Linearizable Read | disk read failure, data corruption alarm, linearizable read loop deadlock, no raft leader, raft loop deadlock, no raft quorum, defrag | 
+| Watch | disk read failure, data corruption alarm, watch loop deadlock, defrag |
+| Write | raft loop deadlock, stalled disk write,  no raft quorum, defrag *(failure to write to stable or memory storage would result in FATAL or panic already) |
+
+In this initial iteration of the probes, we will only focus on the functions of Serializable Read and Linearizable Read. The probing of Watch and Write would merit their own dedicated discussions in the future.
+
+#### Detection
+The table below shows the checks we plan to implement to detect the aforementioned failure modes. This list just reflects the initial implementation, and is not supposed to be exhaustive. We will make the checks to be easily extensible for more checks to be added in the future. 
+
+| Health Check Name | Health Check Group | Related Failure Modes | Health Check Method* |
+| ----------- | ----------- | ----------- | ----------- | 
+| data_corruption | /readyz | data corruption alarm | check for active alarm of AlarmType_CORRUPT. |
+| read_index** | /readyz | no raft leader, raft loop deadlock | check if the server can get ReadIndex. |
+| serializable_read | /readyz, /livez | mvcc read failure | check if a serializable range (limit 1) request returns error, precondition on: defrag is not active. |
+| ... | | | |
+
+*We expect to execute the readyz check every 1s, and livez check every 5s. Any health check within that group should be able to finish below that time scale under normal circumstances. 
+
+**Current health check checks if a linearizable read could finish. We prefer just checking the read index instead of doing a full linearizable read because we expect to execute the ready check every 1s, and a full linearizable read could timeout while the local server is trying to catch up with the applied entries, which is not covered under readiness.
+
+## Implementation Plan
+The following steps would be required to implement this change:
+1. Add two new probes to etcd: a liveness probe and a readiness probe.
+1. Add http handlers that could detect the above failure modes.
+1. Integration and E2E tests the changes with failure mode simulation to ensure that probers work as expected.
+1. Back port the changes to supported versions (3.4 and 3.5).
+1. Update the etcd documentation to reflect the changes.
+
+## Future Discussions/Improvements
+1. There are several remaining topics that are worth more discussion or more work in the future to improve livez/readyz:
+1. Should readyz check include checking writes?
+1. Should readyz check cover performance issues?
+1. What checks can we do to make sure watch is working properly?
+1. In order to catch deadlock in raft loop in livez prober, is there a way to do this without involving external dependencies in multi-node scenario?
```

Comment:
- Format issue.. they should be sub-items of "There are several remaining topics...";

Reply:
done